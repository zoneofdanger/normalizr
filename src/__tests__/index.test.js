// eslint-env jest
import { denormalize, normalize, schema } from '../';

describe('normalize', () => {
  [42, null, undefined, '42', () => {}].forEach((input) => {
    test(`cannot normalize input that == ${input}`, () => {
      expect(() => normalize(input, new schema.Entity('test'))).toThrow();
    });
  });

  test('cannot normalize without a schema', () => {
    expect(() => normalize({})).toThrow();
  });

  test('normalizes entities', () => {
    const mySchema = new schema.Entity('tacos');

    expect(normalize([{ id: 1, type: 'foo' }, { id: 2, type: 'bar' }], [mySchema])).toMatchSnapshot();
  });

  test('normalizes entities with circular references', () => {
    const user = new schema.Entity('users');
    user.define({
      friends: [user]
    });

    const input = { id: 123, friends: [] };
    input.friends.push(input);

    const normalized = {
      entities: {
        users: {
          '123': {
            id: 123,
            friends: [123]
          }
        }
      },
      result: 123
    };

    expect(normalize(input, user)).toMatchSnapshot();
  });

  test('normalizes nested entities', () => {
    const user = new schema.Entity('users');
    const comment = new schema.Entity('comments', {
      user: user
    });
    const article = new schema.Entity('articles', {
      author: user,
      comments: [comment]
    });

    const input = {
      id: '123',
      title: 'A Great Article',
      author: {
        id: '8472',
        name: 'Paul'
      },
      body: 'This article is great.',
      comments: [
        {
          id: 'comment-123-4738',
          comment: 'I like it!',
          user: {
            id: '10293',
            name: 'Jane'
          }
        }
      ]
    };
    expect(normalize(input, article)).toMatchSnapshot();
  });

  test('does not modify the original input', () => {
    const user = new schema.Entity('users');
    const article = new schema.Entity('articles', { author: user });
    const input = Object.freeze({
      id: '123',
      title: 'A Great Article',
      author: Object.freeze({
        id: '8472',
        name: 'Paul'
      })
    });
    expect(() => normalize(input, article)).not.toThrow();
  });

  test('ignores null values', () => {
    const myEntity = new schema.Entity('myentities');
    expect(normalize([null], [myEntity])).toMatchSnapshot();
    expect(normalize([undefined], [myEntity])).toMatchSnapshot();
    expect(normalize([false], [myEntity])).toMatchSnapshot();
  });

  test('can use fully custom entity classes', () => {
    class MyEntity extends schema.Entity {
      schema = {
        children: [new schema.Entity('children')]
      };

      getId(entity, parent, key) {
        return entity.uuid;
      }

      normalize(input, parent, key, visit, addEntity, visitedEntities) {
        const entity = { ...input };
        Object.keys(this.schema).forEach((key) => {
          const schema = this.schema[key];
          entity[key] = visit(input[key], input, key, schema, addEntity, visitedEntities);
        });
        addEntity(this, entity, parent, key);
        return {
          uuid: this.getId(entity),
          schema: this.key
        };
      }
    }

    const mySchema = new MyEntity('food');
    expect(
      normalize(
        {
          uuid: '1234',
          name: 'tacos',
          children: [{ id: 4, name: 'lettuce' }]
        },
        mySchema
      )
    ).toMatchSnapshot();
  });

  test('uses the non-normalized input when getting the ID for an entity', () => {
    const userEntity = new schema.Entity('users');
    const idAttributeFn = jest.fn((nonNormalized, parent, key) => nonNormalized.user.id);
    const recommendation = new schema.Entity(
      'recommendations',
      { user: userEntity },
      {
        idAttribute: idAttributeFn
      }
    );
    expect(normalize({ user: { id: '456' } }, recommendation)).toMatchSnapshot();
    expect(idAttributeFn.mock.calls).toMatchSnapshot();
    expect(recommendation.idAttribute).toBe(idAttributeFn);
  });

  test('passes over pre-normalized values', () => {
    const userEntity = new schema.Entity('users');
    const articleEntity = new schema.Entity('articles', { author: userEntity });

    expect(normalize({ id: '123', title: 'normalizr is great!', author: 1 }, articleEntity)).toMatchSnapshot();
  });

  test('can normalize object without proper object prototype inheritance', () => {
    const test = { id: 1, elements: [] };
    test.elements.push(
      Object.assign(Object.create(null), {
        id: 18,
        name: 'test'
      })
    );

    const testEntity = new schema.Entity('test', {
      elements: [new schema.Entity('elements')]
    });

    expect(() => normalize(test, testEntity)).not.toThrow();
  });
});

describe('denormalize', () => {
  test('cannot denormalize without a schema', () => {
    expect(() => denormalize({})).toThrow();
  });

  test('returns the input if undefined', () => {
    expect(denormalize(undefined, {}, {})).toBeUndefined();
  });

  test('denormalizes entities', () => {
    const mySchema = new schema.Entity('tacos');
    const entities = {
      tacos: {
        1: { id: 1, type: 'foo' },
        2: { id: 2, type: 'bar' }
      }
    };
    expect(denormalize([1, 2], [mySchema], entities)).toMatchSnapshot();
  });

  test('denormalizes nested entities', () => {
    const user = new schema.Entity('users');
    const comment = new schema.Entity('comments', {
      user: user
    });
    const article = new schema.Entity('articles', {
      author: user,
      comments: [comment]
    });

    const entities = {
      articles: {
        '123': {
          author: '8472',
          body: 'This article is great.',
          comments: ['comment-123-4738'],
          id: '123',
          title: 'A Great Article'
        }
      },
      comments: {
        'comment-123-4738': {
          comment: 'I like it!',
          id: 'comment-123-4738',
          user: '10293'
        }
      },
      users: {
        '10293': {
          id: '10293',
          name: 'Jane'
        },
        '8472': {
          id: '8472',
          name: 'Paul'
        }
      }
    };
    expect(denormalize('123', article, entities)).toMatchSnapshot();
  });

  test('does not modify the original entities', () => {
    const user = new schema.Entity('users');
    const article = new schema.Entity('articles', { author: user });
    const entities = Object.freeze({
      articles: Object.freeze({
        '123': Object.freeze({
          id: '123',
          title: 'A Great Article',
          author: '8472'
        })
      }),
      users: Object.freeze({
        '8472': Object.freeze({
          id: '8472',
          name: 'Paul'
        })
      })
    });
    expect(() => denormalize('123', article, entities)).not.toThrow();
  });

  test('denormalizes with function as idAttribute', () => {
    const normalizedData = {
      entities: {
        patrons: {
          '1': { id: '1', guest: null, name: 'Esther' },
          '2': { id: '2', guest: 'guest-2-1', name: 'Tom' }
        },
        guests: { 'guest-2-1': { guest_id: 1 } }
      },
      result: ['1', '2']
    };

    const guestSchema = new schema.Entity(
      'guests',
      {},
      {
        idAttribute: (value, parent, key) => `${key}-${parent.id}-${value.guest_id}`
      }
    );

    const patronsSchema = new schema.Entity('patrons', {
      guest: guestSchema
    });

    expect(denormalize(normalizedData.result, [patronsSchema], normalizedData.entities)).toMatchSnapshot();
  });
});
