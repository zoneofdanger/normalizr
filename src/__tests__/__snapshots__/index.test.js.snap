// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`denormalize denormalizes entities 1`] = `
Array [
  Object {
    "id": 1,
    "type": "foo",
  },
  Object {
    "id": 2,
    "type": "bar",
  },
]
`;

exports[`denormalize denormalizes nested entities 1`] = `
Object {
  "author": Object {
    "id": "8472",
    "name": "Paul",
  },
  "body": "This article is great.",
  "comments": Array [
    Object {
      "comment": "I like it!",
      "id": "comment-123-4738",
      "user": Object {
        "id": "10293",
        "name": "Jane",
      },
    },
  ],
  "id": "123",
  "title": "A Great Article",
}
`;

exports[`denormalize denormalizes with function as idAttribute 1`] = `
Array [
  Object {
    "guest": null,
    "id": "1",
    "name": "Esther",
  },
  Object {
    "guest": Object {
      "guest_id": 1,
    },
    "id": "2",
    "name": "Tom",
  },
]
`;

exports[`normalize can use fully custom entity classes 1`] = `
Object {
  "entities": Object {
    "children": Object {
      "4": Object {
        "id": 4,
        "name": "lettuce",
      },
    },
    "food": Object {
      "1234": Object {
        "children": Array [
          4,
        ],
        "name": "tacos",
        "uuid": "1234",
      },
    },
  },
  "result": Object {
    "schema": "food",
    "uuid": "1234",
  },
}
`;

exports[`normalize ignores null values 1`] = `
Object {
  "entities": Object {},
  "result": Array [
    null,
  ],
}
`;

exports[`normalize ignores null values 2`] = `
Object {
  "entities": Object {},
  "result": Array [
    undefined,
  ],
}
`;

exports[`normalize ignores null values 3`] = `
Object {
  "entities": Object {},
  "result": Array [
    false,
  ],
}
`;

exports[`normalize normalizes entities 1`] = `
Object {
  "entities": Object {
    "tacos": Object {
      "1": Object {
        "id": 1,
        "type": "foo",
      },
      "2": Object {
        "id": 2,
        "type": "bar",
      },
    },
  },
  "result": Array [
    1,
    2,
  ],
}
`;

exports[`normalize normalizes entities with circular references 1`] = `
Object {
  "entities": Object {
    "users": Object {
      "123": Object {
        "friends": Array [
          123,
        ],
        "id": 123,
      },
    },
  },
  "result": 123,
}
`;

exports[`normalize normalizes nested entities 1`] = `
Object {
  "entities": Object {
    "articles": Object {
      "123": Object {
        "author": "8472",
        "body": "This article is great.",
        "comments": Array [
          "comment-123-4738",
        ],
        "id": "123",
        "title": "A Great Article",
      },
    },
    "comments": Object {
      "comment-123-4738": Object {
        "comment": "I like it!",
        "id": "comment-123-4738",
        "user": "10293",
      },
    },
    "users": Object {
      "10293": Object {
        "id": "10293",
        "name": "Jane",
      },
      "8472": Object {
        "id": "8472",
        "name": "Paul",
      },
    },
  },
  "result": "123",
}
`;

exports[`normalize passes over pre-normalized values 1`] = `
Object {
  "entities": Object {
    "articles": Object {
      "123": Object {
        "author": 1,
        "id": "123",
        "title": "normalizr is great!",
      },
    },
  },
  "result": "123",
}
`;

exports[`normalize uses the non-normalized input when getting the ID for an entity 1`] = `
Object {
  "entities": Object {
    "recommendations": Object {
      "456": Object {
        "user": "456",
      },
    },
    "users": Object {
      "456": Object {
        "id": "456",
      },
    },
  },
  "result": "456",
}
`;

exports[`normalize uses the non-normalized input when getting the ID for an entity 2`] = `
Array [
  Array [
    Object {
      "user": Object {
        "id": "456",
      },
    },
    Object {
      "user": Object {
        "id": "456",
      },
    },
    null,
  ],
  Array [
    Object {
      "user": Object {
        "id": "456",
      },
    },
    Object {
      "user": Object {
        "id": "456",
      },
    },
    null,
  ],
]
`;
